### PR TITLE
:sparkles: added client builder

### PR DESCRIPTION
**Pull request by Shahab96**
- The setup script for tests seems to use memcached 1.5.20 and the gitignore needed to be updated so I've added that as well. Tests now use the latest version of memcached (1.6.26 as of the time of this PR)
- While read and write timeouts are set in the client builder, I am unsure if the connection would retain those settings if it gets dropped and a new one is made.

## Summary by CodeRabbit

- **New Features**
	- Enhanced client connections with a `ClientBuilder` for customizable configurations.

- **Refactor**
	- Improved client connection setup process for better performance and usability.

- **Chores**
	- Updated `.gitignore` to reference the latest `memcached` version.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -434,6 +438,93 @@ impl Client {
...
+        self.write_timeout = Some(write_timeout);
+        self
+    }
+
+    pub fn pool_wait_timeout(mut self, pool_wait_timeout: Duration) -> Self {
+        self.pool_wait_timeout = pool_wait_timeout;
+        self
+    }
+
+    #[must_use]
```
Comment (Shahab96) [inline src/client.rs:494]: On second thought I'm wondering if this is truly needed. I don't imagine people would make a connection and just discard it.

**Comment by Shahab96** [comment]
@coderabbitai review

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "memcache"
-version = "0.17.2"
+version = "0.18.0"
```
Comment (aisk) [inline Cargo.toml:3]: Usually we don't bump the version in common changes, we just change it before release.

**Comment by aisk** [comment]
Hi, thank you for contribution, but can you add some basic test case for these newly added codes?

And the coderabbitai is no quite useful now, please feel free to ignore it if you think it's wroung.

**Comment by shadogar** [comment]
> Hi, thank you for contribution, but can you add some basic test case for these newly added codes?
> 
> And the coderabbitai is no quite useful now, please feel free to ignore it if you think it's wroung.

I can add some unit tests to check the connection being created correctly, the connection itself then gets tested by the existing tests though. Would that be good?

**Comment by aisk** [comment]
Yes, please go ahead.

File: src/client.rs [hunk 3]
```diff
@@ -434,20 +438,117 @@ impl Client {
...
+            pool_wait_timeout: Duration::from_secs(30),
+        }
+    }
+
+    pub fn with_max_size(mut self, max_size: u32) -> Self {
+        self.max_size = max_size;
+        self
+    }
+
+    pub fn with_min_idle(mut self, min_idle: u32) -> Self {
```
Comment (aisk) [inline src/client.rs:469]: I think these configurations is for r2d2 pool, but with the name `max_size` or `min_idle` is really confusing. Maybe some user will thinks the `max_size` is for configure the max item size in memcached.

So it will be more clear to add a prefix, like `pool_`.

Comment (Shahab96) [reply src/client.rs:469]: Makes sense, I'll change that and take another look at the rest of the names

**Comment by aisk** [comment]
I think the test code you added is a little bit redundant. I think they are not needed, and what I want to test is to check if the `build` method can build something what we specified and nothing is wroung.

**Comment by Shahab96** [comment]
> I think the test code you added is a little bit redundant. I think they are not needed, and what I want to test is to check if the `build` method can build something what we specified and nothing is wroung.

I'm not sure what you mean by building something and nothing is wrong, do you mean like using the builder and having the resulting connection not be configured the way the caller requested?

**Comment by aisk** [comment]
Yes, I think we should just test the build result is what we specified.

**Comment by Shahab96** [comment]
> Yes, I think we should just test the build result is what we specified.

I'm looking into adding tests for this but I am not seeing a clear path to do so. `Client` does not expose any public params other than the hash function, which we just use the default anyway. None of the settings we provide can be accessed as params of the client. I'm thinking testing this would actually just be testing that r2d2 works properly, since the builder is really only calling that, and so we shouldn't need to do this.

Please let me know if I'm misunderstanding something here.

File: src/client.rs [hunk 4]
```diff
@@ -434,20 +438,115 @@ impl Client {
...
+        let max_size = self.max_size;
+        let min_idle = self.min_idle;
+        let max_lifetime = self.max_lifetime;
+
+        let connections = urls
+            .iter()
+            .filter_map(|url| {
+                let url = Url::parse(url.as_str()).ok()?;
+
+                r2d2::Pool::builder()
```
Comment (Shahab96) [inline src/client.rs:504]: This line is the reason for my comment on unclear path to unit tests for the builder. All we're doing is providing inputs to r2d2 here

File: src/client.rs [hunk 5]
```diff
@@ -434,20 +444,128 @@ impl Client {
...
 #[cfg(test)]
 mod tests {
+    #[test]
+    fn build_client_happy_path() {
+        let client = super::Client::builder("memcache://localhost:12345").build().unwrap();
+        assert!(client.version().unwrap()[0].1 != "");
+    }
+
+    #[test]
+    fn build_client_bad_url() {
```
Comment (Shahab96) [inline src/client.rs:542]: I was able to come up with this, I do think it's much improved from what I previously wrote, as bad urls were just ignored from the list of connections without informing the caller

**Comment by Shahab96** [comment]
Sorry, I'm new to gitmoji, which emotes should have been used for the failing commits?

File: src/client.rs (hunk 4 above)
Comment (Shahab96) [reply src/client.rs:511]: Outdated comment

**Comment by aisk** [comment]
> Sorry, I'm new to gitmoji, which emotes should have been used for the failing commits?

This doesn't matter, I will squash the commits when merge the commit, so only the PR title will be used as commit message.

**Comment by aisk** [comment]
Hi, I think this PR is almost done, except for the comments above, and the test case.

I think we can just add some test cases. In these test cases, we can build a `Client`, with some `with_` methods. And we can check the result, which should contain a vector of r2d2 `Pool` objects, to see if the expected configs were applied.

I'm not sure if these fields are accessble or not, if not, maybe we can found some other ways to make the test, but the goal should be same.

File: src/client.rs [hunk 6]
```diff
@@ -434,20 +444,128 @@ impl Client {
...
+
+        for url in urls.iter() {
+            let url = Url::parse(url.as_str()).map_err(|e| MemcacheError::BadURL(e.to_string()))?;
+
+            connections.push(
+                r2d2::Pool::builder()
+                    .max_size(max_size)
+                    .min_idle(min_idle)
+                    .max_lifetime(max_lifetime)
+                    .connection_timeout(Duration::from_secs(30))
```
[the remark this answers is not in the document]
Comment (Shahab96) [reply src/client.rs:515]: Ah yeah I forgot to add that, thanks!

**Comment by Shahab96** [comment]
> Hi, I think this PR is almost done, except for the comments above, and the test case.
> 
> I think we can just add some test cases. In these test cases, we can build a `Client`, with some `with_` methods. And we can check the result, which should contain a vector of r2d2 `Pool` objects, to see if the expected configs were applied.
> 
> I'm not sure if these fields are accessble or not, if not, maybe we can found some other ways to make the test, but the goal should be same.

Those fields are all private, but we can still test that the builder works

**Comment by Shahab96** [comment]
> Hi, I think this PR is almost done, except for the comments above, and the test case.
> 
> I think we can just add some test cases. In these test cases, we can build a `Client`, with some `with_` methods. And we can check the result, which should contain a vector of r2d2 `Pool` objects, to see if the expected configs were applied.
> 
> I'm not sure if these fields are accessble or not, if not, maybe we can found some other ways to make the test, but the goal should be same.

It looks like there are ways we can test some of the items, like the connection timeout for example, by simulating activity. However since the current client implementation doesn't have these tests either, it's likely something to add to a separate PR, as this one is just for adding the builder.
I'm still not sure exactly how to approach testing the internals, as at that point we're actually testing r2d2 and not this crate. So it might not be appropriate to do that unless we actually end up running simulations.

File: src/client.rs [hunk 7]
```diff
@@ -434,20 +456,164 @@ impl Client {
...
+        assert!(client.version().unwrap()[0].1 != "");
+    }
+
+    #[test]
+    fn build_client_bad_url() {
+        let client = super::Client::builder()
+            .add_server("memcache://localhost:12345:")
+            .build();
+        assert!(client.is_err());
+    }
```
[the remark this answers is not in the document]
Comment (Shahab96) [reply src/client.rs:591]: Yes

[the remark this answers is not in the document]
Comment (Shahab96) [reply src/client.rs:591]: > The tests `build_client_happy_path` and `build_client_bad_url` effectively verify the basic functionality of the `ClientBuilder`. However, considering the discussion about adding more test cases, it would be beneficial to include tests that cover more scenarios, such as invalid configurations (e.g., setting a negative timeout) or edge cases in the connection parameters. This would ensure the robustness of the `ClientBuilder` implementation.
> 
> Would you like me to help draft additional test cases covering a wider range of scenarios for the `ClientBuilder`?

Yes

File: src/client.rs [hunk 8]
```diff
@@ -18,6 +18,24 @@ pub trait Connectable {
     fn get_urls(self) -> Vec<String>;
 }
 
+impl Connectable for (&str, u16) {
+    fn get_urls(self) -> Vec<String> {
+        return vec![format!("{}:{}", self.0, self.1)];
+    }
+}
+
+impl Connectable for &[(&str, u16)] {
```
Comment (Shahab96) [inline src/client.rs:27]: These two should allow consumers to either repeatedly call `add_server` with a single tuple, or with a slice of tuples. That'll take care of host:port specification.

File: src/client.rs [hunk 9]
```diff
@@ -434,20 +444,128 @@ impl Client {
...
+
+impl<C: Connectable> ClientBuilder<C> {
+    pub fn new(target: C) -> Self {
+        ClientBuilder {
+            target,
+            max_size: 1,
+            min_idle: None,
+            max_lifetime: None,
+            read_timeout: None,
+            write_timeout: None,
```
[the remark this answers is not in the document]
Comment (Shahab96) [reply src/client.rs:479]: This goes against the builder pattern, the method names would have `with_` but not field names. If you still want them I can add them anyway though.

File: src/client.rs [hunk 10]
```diff
@@ -434,20 +456,251 @@ impl Client {
...
+        assert!(client.is_ok(), "Should handle custom hash function gracefully");
+    }
+
+    #[test]
+    fn build_client_with_unsupported_protocol() {
+        let client = super::Client::builder()
+            .add_server("unsupported://localhost:12345")
+            .build();
+        assert!(client.is_err(), "Expected error when using an unsupported protocol");
+    }
```
[the remark this answers is not in the document]
Comment (Shahab96) [reply src/client.rs:664]: No

File: src/client.rs [hunk 11]
```diff
@@ -434,20 +456,251 @@ impl Client {
...
+        let client = Client {
+            connections,
+            hash_function: self.hash_function,
+        };
+
+        client.set_read_timeout(self.read_timeout)?;
+        client.set_write_timeout(self.write_timeout)?;
+
+        Ok(client)
+    }
```
[the remark this answers is not in the document]
Comment (Shahab96) [reply src/client.rs:578]: Bad feedback

File: src/client.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (aisk) [reply src/client.rs:479]: Sorry, I made a misleading, I mean the method name without `with_`, like `pool_max_size`, but not the current `max_size`.

File: src/client.rs [hunk 12]
```diff
@@ -434,20 +456,251 @@ impl Client {
...
     #[test]
     fn unix() {
         let client = super::Client::connect("memcache:///tmp/memcached.sock").unwrap();
         assert!(client.version().unwrap()[0].1 != "");
+
+        let client = super::Client::builder()
+            .add_server("memcache:///tmp/memcached.sock")
+            .build()
+            .unwrap();
+        assert!(client.version().unwrap()[0].1 != "");
```
Comment (aisk) [inline src/client.rs:690]: I think this test case and bellow can be removed, the builder logic has been tested by `build_client_with_all_optional_parameters`. Too many codes may leads to hard to maintain the test.

**Review by aisk — COMMENTED** [review]
Thank you for continue the work, I only have two last comments.

File: src/client.rs [hunk 13]
```diff
@@ -434,20 +456,251 @@ impl Client {
...
+    pub fn build(self) -> Result<Client, MemcacheError> {
+        let urls = self.target.get_urls();
+        let max_size = self.max_size;
+        let min_idle = self.min_idle;
+        let max_lifetime = self.max_lifetime;
+        let timeout = self.connection_timeout;
+
+        let mut connections = vec![];
+
+        for url in urls.iter() {
```
[the remark this answers is not in the document]
Comment (Shahab96) [reply src/client.rs:543]: Good catch, I'll add some checks for this!

**Review by aisk — APPROVED** [review]

**Comment by aisk** [comment]
Thank you for your contribution!
